lib/lib-storage/src/Upload.ts

```diff
@@ -132,9 +183,6 @@ export class Upload extends EventEmitter {
   }
 
   async __doMultipartUpload(): Promise<ServiceOutputTypes> {
-    const createMultipartUploadResult = await this.client.send(new CreateMultipartUploadCommand(this.params));
-    this.uploadId = createMultipartUploadResult.UploadId;
-
     // Set up data input chunks.
     const dataFeeder = getChunk(this.params.Body, this.partSize);
```

Comment:
You can check whether it is a "small" body by attempting the first chunk. You can possibly do the trick without touch other parameters in this class. Here's the idea of it:

```TypeScript
  async __doMultipartUpload(): Promise<ServiceOutputTypes> {
    // Set up data input chunks.
    const dataFeeder = getChunk(this.params.Body, this.partSize);

    const firstChunk = await dataFeeder.next();
    if (firstChunk.done) {
      // putObject
      // TODO
    } else {
      // Multipart upload
      const createMultipartUploadResult = await this.client.send(new CreateMultipartUploadCommand(this.params));
      this.uploadId = createMultipartUploadResult.UploadId;
      const rewindDataFeeder = async function* () {
        yield (firstChunk as IteratorYieldResult<RawDataPart>).value;
        for await (const chunk of dataFeeder) {
          yield chunk;
        }
      };
      // Create and start concurrent uploads.
      for (let index = 0; index < this.queueSize; index++) {
        const currentUpload = this.__doConcurrentUpload(rewindDataFeeder());
        this.concurrentUploaders.push(currentUpload);
      }

      // Create and start concurrent uploads.
      await Promise.all(this.concurrentUploaders);
      if (this.abortController.signal.aborted) {
        throw Object.assign(new Error("Upload aborted."), { name: "AbortError" });
      }
    }

    this.uploadedParts.sort((a, b) => a.PartNumber! - b.PartNumber!);
    const completeMultipartUpload = await this.client.send(
      new CompleteMultipartUploadCommand({
        ...this.params,
        UploadId: this.uploadId,
        MultipartUpload: {
          Parts: this.uploadedParts,
        },
      })
    );

    // Add tags to the object after it's completed the upload.
    if (this.tags.length) {
      await this.client.send(
        new PutObjectTaggingCommand({
          ...this.params,
          Tagging: {
            TagSet: this.tags,
          },
        })
      );
    }

    return completeMultipartUpload;
  }
```

Reply:
Yes, it wouldn't be difficult, as I said the only issue is that the code is more complex IMO. Note that once we go down this route we'd also need to manage the lifetime of the iterator, as we'd need to wrap stuff in a try/finally and call `iterable.return()` to make sure that it isn't missed as otherwise some resources might not get released (which is done by the for await loop "for free"). If it's fine I can change the code. I've already implemented something similar to the above locally.